generator/internal/predicate_utils.h

```diff
@@ -0,0 +1,194 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef GOOGLE_CLOUD_CPP_GENERATOR_INTERNAL_PREDICATE_UTILS_H
+#define GOOGLE_CLOUD_CPP_GENERATOR_INTERNAL_PREDICATE_UTILS_H
+
+#include "google/cloud/optional.h"
+#include <google/protobuf/descriptor.h>
+#include <functional>
+#include <vector>
+
+namespace google {
+namespace cloud {
+namespace generator_internal {
+
+/**
+ * Determines if the given method meets the criteria for pagintation.
+ *
+ * https://google.aip.dev/client-libraries/4233
+ */
+bool IsPaginated(google::protobuf::MethodDescriptor const* method);
+
+/**
+ * Detemines if the given method has neither client-side streaming, server-side
+ * streaming, nor bidirectional streaming.
+ */
+bool IsNonStreaming(google::protobuf::MethodDescriptor const* method);
+
+/**
+ * Determines if the given method is a long running operation.
+ */
+bool IsLongrunningOperation(google::protobuf::MethodDescriptor const* method);
+
+/**
+ * Determines if the given method returns `google::protobug::Empty`.
+ */
+bool IsResponseTypeEmpty(google::protobuf::MethodDescriptor const* method);
+
+/**
+ * If method meets pagination criteria, provides paginated field type and field
+ * name.
+ *
+ * https://google.aip.dev/client-libraries/4233
+ */
+google::cloud::optional<std::pair<std::string, std::string>>
+DeterminePagination(google::protobuf::MethodDescriptor const* method);
+
+/**
+ * Returns true if all predicates return true.
+ */
+template <typename T>
+class GenericAll {
+ public:
+  template <typename... Predicates>
+  explicit GenericAll(Predicates... p) : predicates_({p...}) {}
+
+  bool operator()(T const* m) const {
+    for (auto const& p : predicates_) {
+      if (!p(m)) return false;
+    }
+    return true;
+  }
+
+ private:
+  std::vector<std::function<bool(T const* m)>> predicates_;
+};
+
+using All = GenericAll<google::protobuf::MethodDescriptor>;
+
+/**
+ * Returns true if any predicate returns true.
+ * @tparam T
+ */
+template <typename T>
+class GenericAny {
+ public:
+  template <typename... Predicates>
+  explicit GenericAny(Predicates... p) : predicates_({p...}) {}
+
+  bool operator()(T const* m) const {
+    for (auto const& p : predicates_) {
+      if (p(m)) return true;
+    }
+    return false;
+  }
+
+ private:
+  std::vector<std::function<bool(T const* m)>> predicates_;
+};
+
+using Any = GenericAny<google::protobuf::MethodDescriptor>;
+
+/**
+ * Returns true if both predicates return true.
+ */
+template <typename T>
+class GenericAnd {
+ public:
+  GenericAnd(std::function<bool(T const* m)> lhs,
+             std::function<bool(T const* m)> rhs)
+      : lhs_(std::move(lhs)), rhs_(std::move(rhs)) {}
+  bool operator()(T const* m) const { return lhs_(m) && rhs_(m); }
+
+ private:
+  std::function<bool(T const* m)> lhs_;
+  std::function<bool(T const* m)> rhs_;
+};
+
+using And = GenericAnd<google::protobuf::MethodDescriptor>;
```

Comment:
Do we need `And` and `All`? It seems like `And` is just the case where `All` is given two predicates. Maybe just renamed All -> And, and then delete And?

Similar comment about Or and Any.